contrib/pkg/adm/managedns/enable.go

```diff
@@ -157,47 +171,159 @@ func (o *Options) Run(dynClient client.Client, args []string) error {
+	}
+
+	log.Info("Hive is now ready to create clusters with manageDNS=true")
+	return nil
+}
+
+func waitForHiveAdmissionPods(dynClient dynamic.Interface) error {
+	resourceName := "deployments"
+	gvr := appsv1.SchemeGroupVersion.WithResource(resourceName)
+
 	log.Info("waiting for new hiveadmission pods to deploy")
-	selector := map[string]string{"hiveadmission": "true"}
-	for i := 0; i < 24; i++ {
-		hiveAdmissionPods := &corev1.PodList{}
-		err = dynClient.List(context.Background(), hiveAdmissionPods, client.MatchingLabels(selector))
+
+	unstDeployment, err := dynClient.Resource(gvr).Namespace(constants.HiveNamespace).Get(hiveAdmissionDeployment, metav1.GetOptions{})
+	if err != nil {
+		log.WithError(err).Error("failed to get deployment object")
+		return err
+	}
+
+	deployment := &appsv1.Deployment{}
+	if err := scheme.Scheme.Convert(unstDeployment, deployment, nil); err != nil {
+		log.WithError(err).Error("failed to convert unstructed deployment to deployment")
+		return err
+	}
+
+	statusViewer := &kubectl.DeploymentStatusViewer{}
+
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", hiveAdmissionDeployment).String()
+
+	lw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = fieldSelector
+			return dynClient.Resource(gvr).Namespace(constants.HiveNamespace).List(options)
+
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.FieldSelector = fieldSelector
+			return dynClient.Resource(gvr).Namespace(constants.HiveNamespace).Watch(options)
+		},
+	}
+
+	preconditionFunc := func(store cache.Store) (bool, error) {
+		_, exists, err := store.Get(&metav1.ObjectMeta{Namespace: constants.HiveNamespace, Name: hiveAdmissionDeployment})
 		if err != nil {
-			log.WithError(err).Fatal("error listing hiveadmission pods")
-		}
-		podsReady := true
-		for _, p := range hiveAdmissionPods.Items {
-			if !podHasEnvVar(p, "MANAGED_DOMAINS_FILE") {
-				podsReady = false
-				break
-			}
+			return true, err
 		}
-		if !podsReady {
-			log.Info("new hiveadmission pods not ready, sleeping 5s")
-			time.Sleep(5 * time.Second)
+		if !exists {
+			// We need to make sure we see the object in the cache before we start waiting for events
+			// or we would be waiting for the timeout if such object didn't exist.
+			return true, apierrors.NewNotFound(appsv1.Resource(resourceName), hiveAdmissionDeployment)
 		}
+
+		return false, nil
 	}
-	log.Info("Hive is now ready to create clusters with manageDNS=true")
 
-	return nil
+	ctx, cancel := context.WithTimeout(context.Background(), waitTime)
+	defer cancel()
+
+	_, err = clientwatch.UntilWithSync(ctx, lw, &unstructured.Unstructured{}, preconditionFunc, func(e watch.Event) (bool, error) {
+		switch t := e.Type; t {
+		case watch.Added, watch.Modified:
+			_, done, err := statusViewer.Status(e.Object.(runtime.Unstructured), deployment.Generation)
```

Comment:
This conversion from a `runtime.Object` to a `runtime.Unstructured` should have an OK check.